README.md
# NAME

Catalyst::TraitFor::Controller::CAPTCHA - authenticate human by create and validate captcha

# VERSION

version 1.1

# SYNOPSIS

In your controller

    package MyApp::Controller::MyController;
    use Moose;
    use namespace::autoclean;

    BEGIN { extends 'Catalyst::Controller' }
    with 'Catalyst::TraitFor::Controller::CAPTCHA';

    sub captcha_image  : Local :Args(0) {
        my ( $self, $c ) = @_;
        $c->forward('generate_captcha');
    }

    #Now <img src="/captcha_image" /> will have captcha and 
    #<input name="captcha_text" type="text"> will prompt user to enter captcha text 
    #and this should be embed into your form that needs to be validated.

    sub form_post : Local {
        my ($self,$c) = @_;
    
        my $posted_string = $c->req->body_params('captcha_text');    
        if ( $c->forward('validate_captcha',[$posted_string]) ) {
            #Allowed 
        }else {
            #Not Allowed 
        }
    }

    1;

# SUMMARY

This Catalyst::Controller role provides `Private` methods that deal with the generation and validation of captcha.

## CONFIGURATION

In MyApp.pm (or equivalent in config file):

    __PACKAGE__->config->{captcha} = {
       session_name => 'captcha_string',
       #Refer GD::SecurityImage for additonal configuration  
       gd_config => {
           width => 100,
           height => 50,
           lines => 5,
           gd_font => 'giant',
       },
       create => [qw/normal rect/],
       particle => [10],
       out => {force => 'jpeg'}
   };

## DESCRIPTION

This controller's private methods will create and validate captcha.This module is based/altered from [Catalyst::Plugin::Captcha](https://metacpan.org/pod/Catalyst::Plugin::Captcha) because that shouldn't be a Catalyst::Plugin.Now it is a base controller like [Catalyst::TraitFor::Controller::reCAPTCHA](https://metacpan.org/pod/Catalyst::TraitFor::Controller::reCAPTCHA).It uses [GD::SecurityImage](https://metacpan.org/pod/GD::SecurityImage) and requires a session plugins like [Catalyst::Plugin::Session](https://metacpan.org/pod/Catalyst::Plugin::Session).

## METHODS

### generate\_captcha : Private

This will create and respond the captcha.
 

    $c->forward('generate_captcha');

### validate\_captcha : Private

This will validate the given string  against the Captcha image that has been generated earlier.
 

    if ( $c->forward('validate_captcha',[$posted_string]) ) {
      #do something based on the CAPTCHA passing
    }

# SEE ALSO

- [Catalyst::TraitFor::Controller::reCAPTCHA](https://metacpan.org/pod/Catalyst::TraitFor::Controller::reCAPTCHA)
- [Catalyst::Controller](https://metacpan.org/pod/Catalyst::Controller) 
- [Catalyst](https://metacpan.org/pod/Catalyst)
- [GD::SecurityImage](https://metacpan.org/pod/GD::SecurityImage)

# ACKNOWLEDGEMENTS

This module is almost copied from Diego Kuperman [Catalyst::Plugin::Captcha](https://metacpan.org/pod/Catalyst::Plugin::Captcha).

# AUTHOR

Baskar Nallathambi <baskarmusiri@gmail.com>,<baskar@exceleron.com>

# COPYRIGHT AND LICENSE

This is free module.You can do anything to this module under
the same terms as the Perl 5 programming language system itself.


Changes
Version-1.1 Oct 10 2016
-----------------------
* updated build dependencies (issue Id : 118278)
* kwalitee issues fixes



Version-1.0 Oct 7 2016
-----------------------
* initial commit


t/lib/TestApp.pm
package TestApp;
use Moose;
use namespace::autoclean;

use Catalyst::Runtime 5.80;

use Catalyst qw/
    Session
    Session::Store::FastMmap
    Session::State::Cookie
    /;


extends 'Catalyst';

our $VERSION = '0.01';
$VERSION = eval $VERSION;

__PACKAGE__->config->{captcha} = {
    session_name => 'captcha_string',
    gd_config => {
        width => 100,
        height => 50,
        lines => 5,
        gd_font => 'giant',
    },
    create => [qw/normal rect/],
    particle => [10],
    out => {force => 'jpeg'}
};


__PACKAGE__->setup;

1;


lib/Catalyst/TraitFor/Controller/CAPTCHA.pm
package Catalyst::TraitFor::Controller::CAPTCHA;

use Moose::Role;
use MooseX::MethodAttributes::Role;
use namespace::autoclean;

use GD::SecurityImage;
use HTTP::Date;

our $VERSION = '1.1';

use MRO::Compat;

sub generate_captcha : Private  {
    my ($self,$c)     = @_;

    my $conf  = $c->config->{captcha};

    my $new = $conf->{gd_config}        ||= {};
    my $create = $conf->{create}        ||= [];
    my $particle = $conf->{particle}    ||= [];
    my $out = $conf->{out}              ||= {};
    my $sname = $conf->{session_name}   ||= 'captcha_string';

    my $image = GD::SecurityImage->new( %{ $new } );

    $image->random();
    $image->create( @{ $create } );
    $image->particle( @{ $particle } );

    my ( $image_data, $mime_type, $random_string ) = $image->out( %{ $out } );

    #Store the captcha string to session for validation
    $c->session->{ $sname } = $random_string;

    $c->res->headers->expires( time() );
    $c->res->headers->header( 'Last-Modified' => HTTP::Date::time2str );
    $c->res->headers->header( 'Pragma'        => 'no-cache' );
    $c->res->headers->header( 'Cache-Control' => 'no-cache' );

    $c->res->content_type("image/$mime_type");
    $c->res->output($image_data);
}


sub validate_captcha : Private {
    my ($self, $c , $posted_string ) = @_;
    
    my $conf = $c->config->{captcha};
    my $sname = $conf->{session_name}   ||= 'captcha_string';

    my $string = $c->session->{ $sname };

    #Clear the Captcha 
    $c->session->{ $sname } = undef;

    return ( $posted_string && $string && $posted_string eq $string );

}

1;

__END__
=pod

=head1 NAME

Catalyst::TraitFor::Controller::CAPTCHA - authenticate human by create and validate captcha

=head1 VERSION

version 1.1

=head1 SYNOPSIS

In your controller

    package MyApp::Controller::MyController;
    use Moose;
    use namespace::autoclean;

    BEGIN { extends 'Catalyst::Controller' }
    with 'Catalyst::TraitFor::Controller::CAPTCHA';

    sub captcha_image  : Local :Args(0) {
        my ( $self, $c ) = @_;
        $c->forward('generate_captcha');
    }

    #Now <img src="/captcha_image" /> will have captcha and 
    #<input name="captcha_text" type="text"> will prompt user to enter captcha text 
    #and this should be embed into your form that needs to be validated.

    sub form_post : Local {
        my ($self,$c) = @_;
    
        my $posted_string = $c->req->body_params('captcha_text');    

        if ( $c->forward('validate_captcha',[$posted_string]) ) {
            #Allowed 
        }else {
            #Not Allowed 
        }
    }

    1;

=head1 SUMMARY

This Catalyst::Controller role provides C<Private> methods that deal with the generation and validation of captcha.

=head2 CONFIGURATION

In MyApp.pm (or equivalent in config file):

 __PACKAGE__->config->{captcha} = {
    session_name => 'captcha_string',
    #Refer GD::SecurityImage for additonal configuration  
    gd_config => {
        width => 100,
        height => 50,
        lines => 5,
        gd_font => 'giant',
    },
    create => [qw/normal rect/],
    particle => [10],
    out => {force => 'jpeg'}
};

=head2 DESCRIPTION

This controller's private methods will create and validate captcha.This module is based/altered from L<Catalyst::Plugin::Captcha> because that shouldn't be a Catalyst::Plugin.Now it is a base controller like L<Catalyst::TraitFor::Controller::reCAPTCHA>.It uses L<GD::SecurityImage> and requires a session plugins like L<Catalyst::Plugin::Session>.

=head2 METHODS

=head3 generate_captcha : Private

This will create and respond the captcha.
 
 $c->forward('generate_captcha');

=head3 validate_captcha : Private

This will validate the given string  against the Captcha image that has been generated earlier.
 
 if ( $c->forward('validate_captcha',[$posted_string]) ) {
   #do something based on the CAPTCHA passing
 }

=head1 REPOSITORY

L<https://github.com/Virendrabaskar/Catalyst-TraitFor-Controller-CAPTCHA>

=head1 SEE ALSO

=over 4

=item *

L<Catalyst::TraitFor::Controller::reCAPTCHA>

=item *

L<Catalyst::Controller> 

=item *

L<Catalyst>

=item *

L<GD::SecurityImage>

=back

=head1 ACKNOWLEDGEMENTS

This module is almost copied from Diego Kuperman L<Catalyst::Plugin::Captcha>.

=head1 AUTHOR

Baskar Nallathambi <baskarmusiri@gmail.com>,<baskar@exceleron.com>

=head1 COPYRIGHT AND LICENSE

This is free module.You can do anything to this module under
the same terms as the Perl 5 programming language system itself.

=cut


Makefile.PL
use strict;
use warnings;
use ExtUtils::MakeMaker 6.30;


my %WriteMakefileArgs = (
  "ABSTRACT" => "authenticate human by create and validate captcha",
  "AUTHOR" => "Baskar Nallathambi <baskarmusiri\@gmail.com>",
  "BUILD_REQUIRES" => {
    "Catalyst" => 0,
    "Catalyst::Controller" => 0,
    "Catalyst::Plugin::Session" => 0,
    "Catalyst::Runtime" => "5.80",
    "GD::SecurityImage" => 0,
    "FindBin" => 0,
    "Moose" => 0,
    "Test::More" => 0,
    "Test::WWW::Mechanize::Catalyst" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "Catalyst-TraitFor-Controller-CAPTCHA",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "Catalyst::TraitFor::Controller::CAPTCHA",
  "PREREQ_PM" => {
    "Moose::Role" => 0,
    "MooseX::MethodAttributes::Role" => 0,
    "namespace::autoclean" => 0
  },
  "VERSION" => "1.1",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

  WriteMakefile(%WriteMakefileArgs);




